### ISSUE TITLE
Optimization Code

### ISSUE THREAD

**Issue by samcarlos17** [samcarlos17/Machine-Learning-Assisted-Optimization-of-Tunable-Circular-Bragg-Grating-Cavities-#1]
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker

# Machine Learning Modules
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.model_selection import KFold
from sklearn.metrics import r2_score, mean_absolute_error

# Pymoo Optimization Modules
from pymoo.core.problem import Problem
from pymoo.algorithms.moo.nsga2 import NSGA2
from pymoo.optimize import minimize
from pymoo.util.nds.non_dominated_sorting import NonDominatedSorting
from pymoo.operators.crossover.sbx import SBX
from pymoo.operators.mutation.pm import PM

# ==============================================================================
# CONFIGURATION
# ==============================================================================
DATA_FILE = 'cbg_2d_sweep_data.txt'
TARGET_WAVELENGTH = 1330.0  
WAVELENGTH_TOL = 1.0        

# ==============================================================================
# STEP 1: LOAD & PREPARE SWEEP DATASET
# ==============================================================================
print("--- Step 1: Loading and Cleaning Dataset ---")

df = pd.read_csv(DATA_FILE, sep=r'\s+', comment='#', on_bad_lines='skip')
feature_cols = ['R_nm', 'Lambda_nm']
target_cols = ['lam_res_nm', 'Fp_peak', 'eta_res']

for col in feature_cols + target_cols:
    df[col] = pd.to_numeric(df[col], errors='coerce')

df = df.dropna(subset=feature_cols + target_cols).reset_index(drop=True)

X_data = df[feature_cols].values
Y_data = df[target_cols].values

xl = np.min(X_data, axis=0)  
xu = np.max(X_data, axis=0)  

print(f"Successfully loaded {len(df)} valid simulation rows.")

# ==============================================================================
# STEP 2: ML SURROGATE TRAINING & VALIDATION
# ==============================================================================
print("\n--- Step 2: Training ExtraTrees Regressor Surrogate ---")

surrogate = ExtraTreesRegressor(n_estimators=300, max_depth=16, min_samples_split=2, random_state=42, n_jobs=-1)

kf = KFold(n_splits=5, shuffle=True, random_state=42)
r2_scores, mae_scores = [], []

for train_idx, val_idx in kf.split(X_data):
    surrogate.fit(X_data[train_idx], Y_data[train_idx])
    preds = surrogate.predict(X_data[val_idx])
    r2_scores.append(r2_score(Y_data[val_idx], preds, multioutput='raw_values'))
    mae_scores.append(mean_absolute_error(Y_data[val_idx], preds, multioutput='raw_values'))

print("Surrogate Model 5-Fold Validation Performance:")
print(f"  • Resonance Wavelength --> R²: {np.mean(r2_scores, axis=0)[0]:.4f}")
print(f"  • Purcell Peak (Fp)    --> R²: {np.mean(r2_scores, axis=0)[1]:.4f}")
print(f"  • Collection Eff (eta) --> R²: {np.mean(r2_scores, axis=0)[2]:.4f}")

surrogate.fit(X_data, Y_data)

# ==============================================================================
# STEP 3: INITIAL POPULATION SEEDING
# ==============================================================================
print("\n--- Step 3: Extracting Empirical Seeds ---")

valid_mask = np.abs(df['lam_res_nm'].values - TARGET_WAVELENGTH) <= WAVELENGTH_TOL
valid_X = X_data[valid_mask]
valid_Y = Y_data[valid_mask]
pop_size = 50

if len(valid_X) > 0:
    F_valid = np.column_stack([-valid_Y[:, 1], -valid_Y[:, 2]])
    fronts = NonDominatedSorting().do(F_valid)
    
    top_seeds = [valid_X[idx] for front in fronts for idx in front]
    X_seeds = np.array(top_seeds[:20])
else:
    closest_indices = np.argsort(np.abs(df['lam_res_nm'].values - TARGET_WAVELENGTH))[:20]
    X_seeds = X_data[closest_indices]

np.random.seed(42)
X_pad = np.random.uniform(low=xl, high=xu, size=(pop_size - len(X_seeds), len(xl)))
X_gen0 = np.vstack([X_seeds, X_pad])

# ==============================================================================
# STEP 4: FORMULATE NSGA-II PROBLEM
# ==============================================================================
class CBGStructureOptimization(Problem):
    def __init__(self, model, xl, xu, target, tol):
        super().__init__(n_var=2, n_obj=2, n_ieq_constr=1, xl=xl, xu=xu)
        self.model = model
        self.target = target
        self.tol = tol

    def _evaluate(self, X, out, *args, **kwargs):
        preds = self.model.predict(X)
        f1, f2 = -preds[:, 1], -preds[:, 2] # Maximize Fp and Eta
        g1 = np.abs(preds[:, 0] - self.target) - self.tol # Constraint
        out["F"] = np.column_stack([f1, f2])
        out["G"] = np.column_stack([g1])

# ==============================================================================
# STEP 5: EXECUTE NSGA-II SEARCH
# ==============================================================================
print("\n--- Step 4: Executing NSGA-II Search ---")

problem = CBGStructureOptimization(surrogate, xl, xu, TARGET_WAVELENGTH, WAVELENGTH_TOL)
algorithm = NSGA2(pop_size=pop_size, sampling=X_gen0, crossover=SBX(prob=0.9, eta=15), mutation=PM(prob=0.5, eta=20), eliminate_duplicates=True)
res = minimize(problem, algorithm, termination=('n_gen', 150), seed=42, verbose=False)

# ==============================================================================
# STEP 6: EXTRACT RESULTS & PLOT ENGINEER-FRIENDLY MAPPING
# ==============================================================================
print("\n--- Step 5: Processing Optimization Results ---")

opt_X = res.X

if opt_X is None or (isinstance(opt_X, float) and np.isnan(opt_X)) or len(opt_X) == 0:
    print(f"\n[!] Error: NSGA-II failed to find designs within {TARGET_WAVELENGTH} ± {WAVELENGTH_TOL} nm.")
else:
    if opt_X.ndim == 1: opt_X = opt_X.reshape(1, -1)
    opt_preds = surrogate.predict(opt_X)

    results_df = pd.DataFrame({
        'R_nm': opt_X[:, 0],
        'Lambda_nm': opt_X[:, 1],
        'lam_res_nm': opt_preds[:, 0],
        'Fp_peak': opt_preds[:, 1],
        'eta_res': opt_preds[:, 2]
    }).sort_values(by='Fp_peak', ascending=False).reset_index(drop=True)

    print("\n=== FABRICATION LOOKUP TABLE (Top 10 Pareto Optimal Candidates) ===")
    print(results_df.head(10).to_string(index=False, float_format="%.3f"))

    # --- Plotting Subplots ---
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(14, 5.5), dpi=120)

    # Left: Performance Space
    sc1 = ax1.scatter(results_df['eta_res'], results_df['Fp_peak'], 
                      c=results_df['lam_res_nm'], cmap='viridis', s=70, edgecolors='k')
    ax1.set_xlabel('Collection Efficiency ($\eta_{res}$)')
    ax1.set_ylabel('Purcell Factor ($F_p$ peak)')
    ax1.set_title('Performance Map')
    ax1.grid(True, linestyle='--', alpha=0.5)

    # Right: Geometry Space
    sc2 = ax2.scatter(results_df['R_nm'], results_df['Lambda_nm'], 
                      c=results_df['lam_res_nm'], cmap='viridis', s=70, edgecolors='k')
    ax2.set_xlabel('Disk Radius $R$ (nm)')
    ax2.set_ylabel('Grating Period $\Lambda$ (nm)')
    ax2.set_title('Required Geometry Map')
    ax2.grid(True, linestyle='--', alpha=0.5)

    # Disable the +1.33e3 offset on the colorbar so it shows true wavelength values
    cbar = fig.colorbar(sc1, ax=[ax1, ax2], shrink=0.85)
    cbar.ax.yaxis.set_major_formatter(ticker.ScalarFormatter(useOffset=False))
    cbar.set_label('Resonance Wavelength (nm)')

    plt.suptitle(f'NSGA-II CBG Design Map ($\lambda_0$ = {TARGET_WAVELENGTH} $\pm$ {WAVELENGTH_TOL} nm)', y=1.02, fontsize=14)
    plt.show()
